File: java/src/main/java/payroll/GenerateUUIDListener.java
package payroll;

import org.springframework.data.mongodb.core.mapping.event.AbstractMongoEventListener;
import org.springframework.data.mongodb.core.mapping.event.BeforeConvertEvent;
import org.springframework.stereotype.Component;

import java.util.UUID;

@Component
public class GenerateUUIDListener extends AbstractMongoEventListener<Employee> {
 
    @Override
    public void onBeforeConvert(BeforeConvertEvent<Employee> event) {
        Employee employee = event.getSource();
        if (employee.isNew()) {
            employee.setId(UUID.randomUUID());
        }
    }
 
}

File: kotlin/src/main/kotlin/payroll/GenerateUUIDListener.kt
package payroll

import org.springframework.data.mongodb.core.mapping.event.AbstractMongoEventListener
import org.springframework.data.mongodb.core.mapping.event.BeforeConvertEvent
import org.springframework.stereotype.Component
import java.util.*

@Component
class GenerateUUIDListener : AbstractMongoEventListener<Employee?>() {
  override fun onBeforeConvert(event: BeforeConvertEvent<Employee?>) {
    val employee = event.source
    if (employee.isNew) {
      employee.id = UUID.randomUUID()
    }
  }
}
